### DISCUSSION TITLE
Indefinite persistence of generated hands, hand metadata, and solutions

### DISCUSSION THREAD

**Discussion by kdblocher** [kdblocher/bridge#44] [Ideas]
Generated hands' solutions are deterministic. It is also space inexpensive. Therefore we can easily create a distributed database with the hand and its double-dummy solution. We can also index this table, or create views and index them, to get hands that fit certain criteria (based off shape, suit length, hcp, honor placement, etc.).

Doing this makes the analysis of a given path in a bidding system very fast, as a large quantity of hands with their solutions can be fetched (and cached forever).

- Hands can be stored as 104-bit integers. (Thanks @eieioAce)
- Solutions can be stored as 20 5-bit integer (tricks taken), keyed by direction (2 bits) and strain (3 bits).
- Several attributes can be precomputed and stored as a dimension table (e.g. shape = 560 total). (Thanks @sievers2501)
- Tables are optimized for reads.
- Writes are easily added in bulk.

As we have more ideas for this, I can keep this header updated until we settle on a design.

**Comment by sievers2501** [comment]
So I have had some time considering the structure of the main hands table, and have the following thoughts.  

The first issue is whether to store the hand as a binary(13) or a uniqueidentifier.  I don't care so much about binary(13) saving 3 bytes of space.  Nevertheless, I can find no disadvantage to use using a binary column rather than a GUID.  My guess is it is just as easy, if not easier, to have the application send the binary(13) to the DB rather than parsing and padding and constructing the GUID.  Given that, unless there are any objections, I recommend using binary(13) for the hand type.

The second issue is whether a the binary(13) hand, as effectively a random number, makes a good clustering key.  Due to it's randomness, the constant insertion of more and more random rows creates time consuming page splits and fragmented data storage.  The decision is whether to create a identity column for the clustering index, and make the binary(13) the PK.  It isn't often that you would want the clustering column to be different then the PK.  The major issue is that any nonclustered index stores an invisible column lookup to link back to the same table to grab non-included columns in any data retrieval.  What I did not know was whether that invisible column that gets included in every index was the PK or the clustering column.  Turns out it's the clustering column!  AND...if the clustering column isn't unique, SQL will attach an extra 4-byte uniquifier to it in order to ensure a one to one lookup, thus more overhead.  

Therefore, in our case, given the structure of the whole table, and how it's going to be used, I think we do want to take the road more traveled and make the clustering column to be the same as the primary key, as the 104-bit random number is going to be coming along for the ride with pretty much any query and thus any index anyway.  So I wouldn't want to specifically be forced to list the PK column to be included if a different clustering column would be coming along as well on any nonclustered index regardless.  Thus, I recommend making the PK the clustering index.  A long way around to the default conclusion.  

The third (and long term) issue to then think about is page splits.  Any group of entries will be inserted at random and cause fragmentation and page splits.  Defrangmenting the clustering index is a really bad idea, because it causes the next insert to re-split the pages where the next group of data is being inserted.  Given the size of the data that's coming in, in the long run I can measure the width of a row, and set a fill factor to allow the addition of one or two entries for any given page without creating a page split and yet that will still save space if the row is narrow enough.  That will result in faster inserts for tens of thousands of hands.  Then in the really long run, we can defrag once in a blue moon with the optimal fill factor again.  But again, that's only if the row is narrow enough that we care.  And that's WAY down the road.

**Reply by kdblocher** [reply]
The first issue is whether to store the hand as a binary(13) or a uniqueidentifier. I don't care so much about binary(13) saving 3 bytes of space. Nevertheless, I can find no disadvantage to use using a binary column rather than a GUID. My guess is it is just as easy, if not easier, to have the application send the binary(13) to the DB rather than parsing and padding and constructing the GUID. Given that, unless there are any objections, I recommend using binary(13) for the hand type.

The only point I can think for this at the moment:
1. Does the database engine know how to optimize queries against `binary(13)`? (easy to check, look at the execution plan)
2. If we want to embed additional data in the PK, we have room with a GUID.
3. We have to deal with the integer encoding ourselves.

> The second issue is whether a the binary(13) hand, as effectively a random number, makes a good clustering key. Due to it's randomness, the constant insertion of more and more random rows creates time consuming page splits and fragmented data storage. The decision is whether to create a identity column for the clustering index, and make the binary(13) the PK. It isn't often that you would want the clustering column to be different then the PK. The major issue is that any nonclustered index stores an invisible column lookup to link back to the same table to grab non-included columns in any data retrieval. What I did not know was whether that invisible column that gets included in every index was the PK or the clustering column. Turns out it's the clustering column! AND...if the clustering column isn't unique, SQL will attach an extra 4-byte uniquifier to it in order to ensure a one to one lookup, thus more overhead.
> 
> Therefore, in our case, given the structure of the whole table, and how it's going to be used, I think we do want to take the road more traveled and make the clustering column to be the same as the primary key, as the 104-bit random number is going to be coming along for the ride with pretty much any query and thus any index anyway. So I wouldn't want to specifically be forced to list the PK column to be included if a different clustering column would be coming along as well on any nonclustered index regardless. Thus, I recommend making the PK the clustering index. A long way around to the default conclusion.

I agree with all of this

> The third (and long term) issue to then think about is page splits. Any group of entries will be inserted at random and cause fragmentation and page splits. Defrangmenting the clustering index is a really bad idea, because it causes the next insert to re-split the pages where the next group of data is being inserted. Given the size of the data that's coming in, in the long run I can measure the width of a row, and set a fill factor to allow the addition of one or two entries for any given page without creating a page split and yet that will still save space if the row is narrow enough. That will result in faster inserts for tens of thousands of hands. Then in the really long run, we can defrag once in a blue moon with the optimal fill factor again. But again, that's only if the row is narrow enough that we care. And that's WAY down the road.

Consider that if we are storing 1B records are more, we are probably doing so in a distributed database, which we can shard if necessary, so this is much less of an issue than if we were forced to have all records available on a single server. So we may not ever have to deal with this as being a big issue.

**Comment by kdblocher** [comment]
To save the solutions for the hand, we have to save 20 5-bit integers (5 strains x 4 directions) and 16 scores per hand (with a score table, we can index all possible scores and the key is 1 byte each) (4 directions * 4 vulnerabilities), not counting additional hand metadata. Our key is fat, so denormalization is tempting. @sievers2501, What do you think? Can you check to see how SQL will help us here?

**Comment by sievers2501** [comment]
1)  Yes, the query execution engine can deal with binary(13)
2)  We shouldn't be encoding a separate piece of data inside the PK.  Having multiple data elements inside the same column is usually a terrible idead.
3)  Regardless of what we choose, the binary encoding will be happening client side.

**Comment by sievers2501** [comment]
I have been thinking about this, and given that our database is more of an OLAP DB, I do think we don't want to over normalize.  Given the 20 tinyint values for tricks made per strain per dealer from DDS, and 16 smallints per deal per vulnerability for par score, we are only talking about 52 bytes of data.  Moreover, there is potential drawback to normalize it because of the following consideration.   

For hand shapes, we know ahead of time that there 560, and only 560, hand shapes.  We can pre-define them and pre-populate the shape table that holds the hand shape.  This data can be used as the shape value for ANY of the four hands.  However, consider the case of par scores for vulnerability only.  We have no idea ahead of time how many combinations of the scores would exist.  We could guess, but that means the application, when adding data, would first need to check to see whether the set of four values existed in our "par score" table.  If it didn't, it would need to add that.  This means multiple round trips, and distraction from the fast inserts that are more important.   

For these reasons, I am leaving the shape table, as I think that's a great normalization technique for hand pattern, but everything else I am gong to add to the 'deals" table as individual columns.  It will make the table "feel" wide, but really it's still quite narrow.  Probably 100 byes or less when it's all said and done, which will still be 80 records per disk page.
